### PR TITLE
[FLINK-24538][runtime][tests] Fix race condition when offering information to leader event queue in TestingRetrievalBase

### PR DESCRIPTION
**Pull request by autophagy**
## What is the purpose of the change

There currently exists a race condition during `ZooKeeperLeaderElectionTest.testLeaderShouldBeCorrectedWhenOverwritten`, which is caused by `TestingRetrievalBase.offerToLeaderQueue` assigning a value to the base's leader, which happens because of multiple threads offering to the leader queue and this assignment is not synced.

This assignment is not necessary as `waitForNewLeader` and `waitForEmptyLeaderInformation` already assign the leader by polling the blocking queue, which `offerToLeaderQueue` adds to.

This NPE occurs consistently every 20-30 retries of the test. With this fix, I am unable to reproduce the NPE across 1000+ retries.

## Brief change log

  - *Fixed race condition caused by unsynchronized leadership assignment in TestingRetrievalBase*

## Verifying this change

This change added tests and can be verified as follows:

*(example:)*
  - *Running the  `ZooKeeperLeaderElectionTest.testLeaderShouldBeCorrectedWhenOverwritten` test until failure.*
  
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/SettableLeaderRetrievalServiceTest.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
 public class SettableLeaderRetrievalServiceTest extends TestLogger {
 
     private SettableLeaderRetrievalService settableLeaderRetrievalService;
+    private static final long TIMEOUT = 50L;
```
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/SettableLeaderRetrievalServiceTest.java:35]:
```suggestion
    private static final long TIMEOUT_IN_MILLIS = 50L;
```
It's always a good habit to add the unit to variables like that.

Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/SettableLeaderRetrievalServiceTest.java:35]:
```suggestion
    private static final long TIMEOUT_IN_MILLIS = 50L;
```
It's always a good habit to add the unit to variables like that. Alternatively, you could use Duration here and do initialize it with `Duration.ofHours()` to inialize it. In the test, you can then do `TIMEOUT.toMillis()`. That makes it clearer to the reader.

Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/SettableLeaderRetrievalServiceTest.java:35]: On another note: Having low timeouts increases the risk of having flaky tests. I'm wondering whether we should go to something higher like 1 hour.

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks @autophagy . I found one other test where I was wondering whether we should add the wait call as well: `KubernetesLeaderRetrievalDriverTest.testKubernetesLeaderRetrievalOnModifiedWithEmpty`. Did you check that one?

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/SettableLeaderRetrievalServiceTest.java (hunk 1 above)
Comment (autophagy) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/SettableLeaderRetrievalServiceTest.java:35]: Thanks for the suggestion - I believe a few other tests are also using a timeout like this, i'll bring it in line with more what they're doing.

**Comment by autophagy** [comment]
@XComp I did check that one - I don't believe that a check is strictly necessary for `testKubernetesLeaderRetrievalOnModifiedWithEmpty`, but I will add a `waitForEmptyLeaderInformation` call just for consistency's sake.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/SettableLeaderRetrievalServiceTest.java (hunk 1 above)
Comment (autophagy) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/SettableLeaderRetrievalServiceTest.java:35]: Actually, I prefer your suggestion of using a `Duration.ofHours` for clarity. Will push up a change.

**Review by XComp — APPROVED** [review]
LGTM 👍

**Comment by XComp** [comment]
Test failure is independent of the changes applied in this PR: FLINK-25442

**Pull request merged**